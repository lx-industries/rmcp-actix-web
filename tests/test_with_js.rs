use rmcp::transport::streamable_http_server::session::local::LocalSessionManager;
use rmcp_actix_web::{SseService, StreamableHttpService};
use std::sync::Arc;
use tracing_subscriber::{layer::SubscriberExt, util::SubscriberInitExt};
mod common;
use common::calculator::Calculator;

const SSE_BIND_ADDRESS: &str = "127.0.0.1:8000";
const STREAMABLE_HTTP_BIND_ADDRESS: &str = "127.0.0.1:8001";

#[actix_web::test]
async fn test_with_js_client() -> anyhow::Result<()> {
    let _ = tracing_subscriber::registry()
        .with(
            tracing_subscriber::EnvFilter::try_from_default_env()
                .unwrap_or_else(|_| "debug".to_string().into()),
        )
        .with(tracing_subscriber::fmt::layer())
        .try_init();
    tokio::process::Command::new("npm")
        .arg("install")
        .current_dir("tests/test_with_js")
        .spawn()?
        .wait()
        .await?;

    // Create SSE service using builder pattern
    let sse_service = SseService::builder()
        .service_factory(Arc::new(|| Ok(Calculator::new())))
        .build();

    // Start HTTP server with SSE service
    let server = actix_web::HttpServer::new(move || {
        actix_web::App::new()
            .wrap(actix_web::middleware::Logger::default())
            .service(sse_service.clone().scope())
    })
    .bind(SSE_BIND_ADDRESS)?
    .run();

    let server_handle = server.handle();
    let server_task = tokio::spawn(server);

    // Give the server a moment to start
    tokio::time::sleep(tokio::time::Duration::from_millis(500)).await;

    let output = tokio::process::Command::new("node")
        .arg("tests/test_with_js/client.js")
        .output()
        .await?;
    assert!(output.status.success());

    // Capture and validate the actual MCP responses
    let stdout = String::from_utf8(output.stdout)?;
    let mut responses: Vec<serde_json::Value> = stdout
        .lines()
        .filter(|line| !line.is_empty())
        .map(serde_json::from_str)
        .collect::<Result<Vec<_>, _>>()?;

    // Sort arrays for deterministic snapshots (preserve_order handles object properties)
    for response in &mut responses {
        if let Some(tools) = response.get_mut("tools").and_then(|t| t.as_array_mut()) {
            tools.sort_by(|a, b| {
                let name_a = a.get("name").and_then(|n| n.as_str()).unwrap_or("");
                let name_b = b.get("name").and_then(|n| n.as_str()).unwrap_or("");
                name_a.cmp(name_b)
            });
        }
    }

    insta::assert_json_snapshot!("js_sse_client_responses", responses);

    // Shutdown the server
    server_handle.stop(true).await;
    let _ = server_task.await;
    Ok(())
}

#[actix_web::test]
async fn test_with_js_streamable_http_client() -> anyhow::Result<()> {
    let _ = tracing_subscriber::registry()
        .with(
            tracing_subscriber::EnvFilter::try_from_default_env()
                .unwrap_or_else(|_| "debug".to_string().into()),
        )
        .with(tracing_subscriber::fmt::layer())
        .try_init();
    tokio::process::Command::new("npm")
        .arg("install")
        .current_dir("tests/test_with_js")
        .spawn()?
        .wait()
        .await?;

    let http_service = Arc::new(
        StreamableHttpService::builder()
            .service_factory(Arc::new(|| Ok(Calculator::new())))
            .session_manager(Arc::new(LocalSessionManager::default()))
            .stateful_mode(true)
            .build(),
    );

    let server = actix_web::HttpServer::new(move || {
        actix_web::App::new()
            .wrap(actix_web::middleware::Logger::default())
            .service(actix_web::web::scope("/mcp").service(http_service.clone().scope()))
    })
    .bind(STREAMABLE_HTTP_BIND_ADDRESS)?
    .run();

    let server_handle = server.handle();
    let server_task = tokio::spawn(server);

    // Give the server a moment to start
    tokio::time::sleep(tokio::time::Duration::from_millis(500)).await;

    let output = tokio::process::Command::new("node")
        .arg("tests/test_with_js/streamable_client.js")
        .output()
        .await?;
    assert!(output.status.success());

    // Capture and validate the actual MCP responses
    let stdout = String::from_utf8(output.stdout)?;
    let mut responses: Vec<serde_json::Value> = stdout
        .lines()
        .filter(|line| !line.is_empty())
        .map(serde_json::from_str)
        .collect::<Result<Vec<_>, _>>()?;

    // Sort arrays for deterministic snapshots (preserve_order handles object properties)
    for response in &mut responses {
        if let Some(tools) = response.get_mut("tools").and_then(|t| t.as_array_mut()) {
            tools.sort_by(|a, b| {
                let name_a = a.get("name").and_then(|n| n.as_str()).unwrap_or("");
                let name_b = b.get("name").and_then(|n| n.as_str()).unwrap_or("");
                name_a.cmp(name_b)
            });
        }
    }

    insta::assert_json_snapshot!("js_streamable_http_client_responses", responses);

    server_handle.stop(true).await;
    let _ = server_task.await;
    Ok(())
}
