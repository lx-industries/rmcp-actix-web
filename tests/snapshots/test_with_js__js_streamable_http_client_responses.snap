---
source: tests/test_with_js.rs
expression: responses
---
[
  {
    "tools": [
      {
        "name": "sub",
        "description": "Calculate the sub of two numbers",
        "inputSchema": {
          "type": "object",
          "properties": {
            "a": {
              "description": "the left hand side number",
              "type": "integer",
              "format": "int32"
            },
            "b": {
              "description": "the right hand side number",
              "type": "integer",
              "format": "int32"
            }
          },
          "required": [
            "a",
            "b"
          ],
          "$schema": "http://json-schema.org/draft-07/schema#",
          "title": "SubRequest",
          "description": "Request structure for the subtraction operation.\n\nSimilar to SumRequest but for subtraction, showing consistent\nparameter definition patterns."
        }
      },
      {
        "name": "sum",
        "description": "Calculate the sum of two numbers",
        "inputSchema": {
          "type": "object",
          "properties": {
            "a": {
              "description": "the left hand side number",
              "type": "integer",
              "format": "int32"
            },
            "b": {
              "description": "The right hand side number to add",
              "type": "integer",
              "format": "int32"
            }
          },
          "required": [
            "a",
            "b"
          ],
          "$schema": "http://json-schema.org/draft-07/schema#",
          "title": "SumRequest",
          "description": "Request structure for the sum operation.\n\nDemonstrates how to define typed parameters for MCP tools\nwith JSON schema generation support."
        }
      }
    ]
  },
  {
    "resources": []
  },
  {
    "resourceTemplates": []
  },
  {
    "prompts": []
  }
]
